add method to compute average treatment effects

apoorvalal opened the pull request:
Thanks for the nice work on this package; it addresses a lot of pain-points I have with `econML`. An ATE estimator method was missing (possibly because that might not be the most obvious intended use case for the package, but the rest of the API / package ergonomics are nice enough that I decided to use it for that as well), so I added it (only to `drlearner`, since [the variance-of-pseudooutcome construction for standard error is justified in that case](https://stanford.edu/~swager/stats361.pdf#page=20.22) but not for T-learner etc). 

I've also added a notebook in the `docs/examples` directory that verifies that estimates line up with what one gets from `econML` and `doubleML` on a simulated dataset.

 

# Checklist

- [x] Added a `CHANGELOG.rst` entry

apoorvalal commented:
build failures likely due to use of external packages in the new notebook. Should I add them as dev dependencies (or I can install them inside the notebook; unsure if that conforms to your style guide).

kklein commented:
@apoorvalal Thanks a lot for your kind words and effort! :)

I'll get to your PR asap - hopefully on the weekend!

kklein commented on docs/examples/example_estimating_ates.ipynb:
```diff
@@ -0,0 +1,508 @@
+{
```
Thanks a lot for adding an example! :)
For it to be shown properly in the docs, you'll have to add an entry for it in the corresponding index:
https://github.com/Quantco/metalearners/blob/main/docs/examples/index.rst

kklein commented on CHANGELOG.rst:
```diff
@@ -7,6 +7,15 @@
+0.9.1 (2024-07-xx)
+------------------
+
+**New features**
```
Since 0.9.0 hasn't been released yet, you can add this entry to 0.9.0.

kklein commented on metalearners/drlearner.py:
```diff
@@ -331,6 +331,35 @@ def evaluate(
+            )
+        treatment_effect = gamma_matrix.mean(axis=0)
+        standard_error = gamma_matrix.std(axis=0) / np.sqrt(len(X))
+        return np.c_[treatment_effect, standard_error]
```
Your implementation makes a great impression on me and sure looks slick. Would you mind adding a small test for this functionality to https://github.com/Quantco/metalearners/blob/main/tests/test_drlearner.py?

We could compare the returned values against some simple mock data.

Moreover, I think it could be very useful to make sure everything works as expected:
- with more than two variants
- with classification outcomes

If one of those cases doesn't work with the current implementation, it'd be good if we explicitly checked for that and added it to the doc string.

kklein commented on metalearners/drlearner.py:
```diff
@@ -331,6 +331,35 @@ def evaluate(
+        y: Vector,
+        w: Vector,
+    ) -> np.ndarray:
+        """Compute Average Treatment Effect (ATE) for each treatment variant using the Augmented IPW estimator (Robins et al 1994). Does not require fitting a second-stage treatment model: it uses the pseudo-outcome alone and computes the average and SE. Can be used following the `fit_all_nuisance` method.
```
```suggestion
        """Compute Average Treatment Effect (ATE) for each treatment variant using the Augmented IPW estimator (Robins et al 1994). Does not require fitting a second-stage treatment model: it uses the pseudo-outcome alone and computes the average and SE. Can be used following the :meth:`~metalearners.drlearner.DRLearner.fit_all_nuisance` method.
```

This way a user can get a cross-reference in the docs. :)

kklein reviewed (commented):
Looking great - just some minor comments. :)

kklein commented:
A simple `pixi update` should suffice to get rid of the merge conflict. :)

apoorvalal commented:
addressed most of your (great) suggestions in [this](https://github.com/Quantco/metalearners/pull/73/commits/246f7e8a25a732b86d7a51b353a3a19a55d0a517) commit. I think the conflicts are likely straightforward to resolve (amended a the drlearner test file, added the feature desc to changelog, and the usual pixi lock conflict

apoorvalal commented:
Ready to merge pending (opaque) pixi trouble. 
I added you as a collaborator on my PR so you can directly commit to my `ate` branch rather than to have to send me PRs.

kklein reviewed (approved):
LGTM! Thanks a lot for your contribution @apoorvalal ! :)

I'll create a new release for version 0.9.0 soon.
